Add Windows support and cross-platform line ending compatibility

Kabxx opened the pull request:
## Overview
This pull request consists of two main enhancements to the `lazycsv` library:

1. **Windows Platform Support**: Added proper implementation for the Windows platform to enable memory-mapped file reading on Windows systems.

2. **Line Ending Compatibility**: Enhanced the parser to correctly handle different line ending formats (CRLF/LF) across operating systems.

## Changes

### Windows Support
- Implemented Windows-specific code for memory-mapped files using the Windows API (`CreateFileA`, `CreateFileMappingA`, `MapViewOfFile`)
- Added proper error handling and resource management for Windows handles
- Ensured clean resource deallocation in move operations and destructors

### Line Ending Compatibility
- Added detection and handling for CRLF line endings in the `cell` constructor
- Automatically removes trailing `\r` characters when detected
- Updated tests to use `std::filesystem` for file size verification instead of hardcoded values

## Testing
All tests pass successfully on both Windows and Unix-based systems. 

Windows:
![windows](https://github.com/user-attachments/assets/ee2bcdf1-e4bf-40a4-929b-d8e6a9559a81)

Linux:
![linux](https://github.com/user-attachments/assets/24eb80f9-0966-4e30-a63b-ff31de6b327e)



## Benefits
- Expands usability to Windows environments
- Improves cross-platform compatibility
- Makes tests more reliable and portable
- Files created on any platform can now be correctly read on any other platform

## Implementation Details
The Windows implementation follows the same pattern as the POSIX implementation but uses Windows-specific APIs. The line ending handling is minimal and targeted - adjusting the `end_` pointer in the cell constructor when a `\r` character is detected before a newline.

kabxx commented:
This PR addresses the following issues:
- Implements Windows platform support, References #12
- Handles CRLF line endings properly, References #9

ashtum commented on include/lazycsv.hpp:
```diff
@@ -9,10 +9,16 @@
+#include <windows.h>
+#endif
+
+#if defined(__unix__) || defined(__unix) || defined(__linux__) || defined(__APPLE__)
```
Can we use `#else` instead? If someone uses this code on a platform other than Windows or Unix, they'll get a compile error anyway.

ashtum commented on include/lazycsv.hpp:
```diff
@@ -180,11 +186,77 @@ class mmap_source
+#endif
+#if defined(__unix__) || defined(__unix) || defined(__linux__) || defined(__APPLE__)
     int fd_;
+#endif
```
If we reuse `fd_` there will be no need for extra code for handling `file_handle_` in move ctro and assignment, something like this: 
```
#if defined(_WIN32)
    HANDLE fd_{ INVALID_HANDLE_VALUE };
    HANDLE mapping_handle_{ NULL };
#else
    int fd_;
#endif
```

ashtum commented on include/lazycsv.hpp:
```diff
@@ -246,8 +335,17 @@ class mmap_source
+            if (mapping_handle_ != NULL)
+                CloseHandle(mapping_handle_);
+            if (file_handle_ != INVALID_HANDLE_VALUE)
+                CloseHandle(file_handle_);
```
Is it possible for `mapping_handle_` and `file_handle_` to be invalid while `data_ != nullptr`?

ashtum commented on include/lazycsv.hpp:
```diff
@@ -314,6 +412,9 @@ class parser
             : begin_(escape_leading_quote(begin, end))
             , end_(escape_trailing_quote(begin, end))
         {
+            // Handle CRLF line endings by adjusting end_
+            if (end_ > begin_ && *(end_ - 1) == '\r')
+                --end_;
```
I think this should go into the `row` constructor because a quoted cell can contain `\n`. It will also be more efficient, as it only needs to be checked once per row.

ashtum commented on include/lazycsv.hpp:
```diff
@@ -219,16 +292,32 @@ class mmap_source
     {
         other.data_ = nullptr;
+#if defined(_WIN32)
+        other.file_handle_ = INVALID_HANDLE_VALUE;
+        other.mapping_handle_ = NULL;
```
No need for these assignments, checking `data_` in the dtor should be enough.

ashtum reviewed (commented):
This is brilliant!
I just added a few minor suggestions.

kabxx commented:
Thank you for reviewing my PR. I've also addressed some of the issues you previously mentioned.

I'm not sure if using `#else` in this case might lead to potential compile errors on less common platforms where headers may not be available.

Some of the failing CI tests are on older compilers (g++-7/8, clang-6.0/7/8) and may be failing due to limited C++17 support (filesystem) in these versions. All tests pass successfully on more recent compiler versions (g++-9+ and clang-9+).

ashtum commented:
> Some of the failing CI tests are on older compilers (g++-7/8, clang-6.0/7/8) and may be failing due to limited C++17 support (filesystem) in these versions. All tests pass successfully on more recent compiler versions (g++-9+ and clang-9+).

Yes, the `std::filesystem` implementation is buggy in older versions of libstdc++, and I believe the segmentation fault here is caused by the same issue: https://github.com/ashtum/lazycsv/actions/runs/14834315659/job/41642701199?pr=15#step:7:54

Given this limitation, I believe using a fixed literal for the file size is acceptable.

kabxx commented:
That's a great point — however, when line endings are converted from LF to CRLF (as Git does when checking out code on Windows), the file size can exceed 47 bytes, which can cause the test to fail. To avoid such inconsistencies, perhaps we could remove the test, use `std::filesystem`, or pin the CI to a specific version.

ashtum commented:
> That's a great point — however, when line endings are converted from LF to CRLF (as Git does when checking out code on Windows), the file size can exceed 47 bytes, which can cause the test to fail. To avoid such inconsistencies, perhaps we could remove the test, use `std::filesystem`, or pin the CI to a specific version.

The contents of `basic.csv` is fixed on all platforms so it shouldn't be an issue on Windows.
I'll add more tests to cover `CRLF` test, also extending the CI to run tests on windows machines.

kabxx commented:
> The contents of basic.csv is fixed on all platforms so it shouldn't be an issue on Windows.

Thanks! I'll revert the test file as suggested. 

Just a quick note: when cloning the repository on Windows, Git may convert `LF` to `CRLF` in `basic.csv`, which increases its size beyond 47 bytes and can lead to test failures. We should consider adding a warning for users about this behavior.

ashtum commented:
> Just a quick note: when cloning the repository on Windows, Git may convert `LF` to `CRLF` in `basic.csv`, which increases its size beyond 47 bytes and can lead to test failures. We should consider adding a warning for users about this behavior.

Yeah, I didn't consider that. It would be an issue when I add the Windows platform to the CI as well. For now, I think I'll just use a different value depending on the platform.
